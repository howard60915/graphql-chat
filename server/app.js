import graphqlServer from "./graphql";

const PORT = process.env.NODE_PORT || process.env.PORT || 8080;

const options = {
  port: PORT,
  endpoint: "/graphql",
  playground: "/playground"
};

graphqlServer.start(options, ({ port }) => {
  // eslint-disable-next-line no-console
  console.log(`> Ready on http://localhost:${port}`);
});
