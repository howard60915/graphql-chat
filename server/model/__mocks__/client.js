import { Client } from "pg";
import { mock } from "graphql-shortcake";
import knex from "knex";
import config from "../../../knexfile";

const { POSTGRES_HOST } = process.env;

const client = jest.requireActual("../client").default;

const db = knex(config);

const mockClient = mock(client);

mockClient.reset = async name => {
  const pgClient = new Client({
    host: POSTGRES_HOST || "127.0.0.1",
    user: "postgres",
    password: null,
    database: "postgres"
  });
  await pgClient.connect();
  await pgClient.query(`DROP DATABASE IF EXISTS ${name}`);
  await pgClient.query(`CREATE DATABASE ${name}`);
  await pgClient.end();
  await db.migrate.latest();
};

export default mockClient;
