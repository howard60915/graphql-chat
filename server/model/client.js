import { Client } from "graphql-shortcake";
import config from "../../knexfile";

export default new Client({ connectionString: config.connection });
