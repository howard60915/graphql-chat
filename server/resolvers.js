import UsersQuery from "./query/UsersQuery";
import RegisterByEmailMutation from "./mutation/RegisterByEmailMutation";
import UpdateUserMutation from "./mutation/UpdateUserMutation";
import DelUserMutation from "./mutation/DelUserMutation";
import AddMessageMutation from "./mutation/AddMessageMutation";

const Query = {
  users: UsersQuery
};

const Mutation = {
  registerByEmail: RegisterByEmailMutation,
  updateUser: UpdateUserMutation,
  delUser: DelUserMutation,
  addMessage: AddMessageMutation
};

export default { Query, Mutation };
