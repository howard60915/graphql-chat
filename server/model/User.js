import { GlobalModel, Hash } from "graphql-shortcake";
import client from "./client";

export default class User extends GlobalModel {
  static client = client;

  static hasOperator = true;

  static operator = () => User;

  static tableName = "user";

  static columns = {
    email: { type: String },
    password: { type: Hash, paths: ["archive"] },
    isAdmin: { type: Boolean, paths: ["archive"] }
  };

  async checkWritability(user, ThrowError) {
    const result = this.id === user.id || user.isAdmin;
    if (!result) throw new ThrowError();
    return result;
  }
}
