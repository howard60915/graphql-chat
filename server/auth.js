import { NotFoundError } from "graphql-schema-sushi";
import { AuthenticationError } from "apollo-server-errors";
import { JsonWebToken } from "nextjs-apollo/server";
import User from "./model/User";
import Certificate from "./model/Certificate";

const onSignIn = async data => {
  const { email, password, correlationId } = data;

  const { fetcher } = User;
  const user = await fetcher.where("email", email).fetch(NotFoundError);

  if (!(password && user.password.verify(password))) {
    throw new AuthenticationError("AuthenticationError");
  }

  return { user: user.previous, correlationId };
};

const onSignOut = () => ({});

const onRenew = async data => {
  const { user, correlationId } = data;

  const renewUser = await User.load(user.id, NotFoundError);

  return { user: renewUser.previous, correlationId };
};

const toUser = ({ user }) => {
  if (!user.id) throw new AuthenticationError();
  return User.forge(user);
};

const jwt = new JsonWebToken(Certificate, {
  onSignIn,
  onSignOut,
  onRenew,
  toUser
});

export default jwt;
