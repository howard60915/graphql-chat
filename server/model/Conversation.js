import Model from "./Model";
import User from "./User";

export default class Conversation extends Model {
  static tableName = "conversation";

  static columns = {
    user: { type: User, name: "userId" },
    talker: { type: User, name: "talkerId" }
  };
}
