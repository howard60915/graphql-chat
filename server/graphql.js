import { GraphQLServer } from "graphql-yoga";
import auth from "./auth";

const server = new GraphQLServer({
  typeDefs: `
    type Query {
      hello: String
    }
  `,
  resolvers: {
    Query: {
      hello: () => "hello"
    }
  }
});

auth.applyMiddleware(server.express);

export default server;
