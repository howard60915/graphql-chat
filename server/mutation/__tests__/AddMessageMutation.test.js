import client from "../../model/client";
import User from "../../model/User";
import Conversation from "../../model/Conversation";
import server, { mutation } from "../../__tests__";

const { addMessage } = mutation;

describe("AddMessageMutation", () => {
  const conversationId = new Conversation(
    "00000000-0000-0000-0000-000000000015"
  ).id;
  const content = "hello world";
  const input = { conversationId, content };

  afterAll(() => new Promise(() => server.close()));

  it("successfully add message", async () => {
    client.mockReturnValueOnce([
      {
        id: "00000000-0000-0000-0000-000000000015",
        user_id: "00000000-0000-0000-0000-000000000010",
        conversation_id: "00000000-0000-0000-0000-000000000015"
      }
    ]);
    client.mockReturnValueOnce([
      {
        id: "00000000-0000-0000-0000-000000000015"
      }
    ]);
    client.mockReturnValueOnce([
      {
        id: "00000000-0000-0000-0000-000000000020",
        content,
        conversation_id: "00000000-0000-0000-0000-000000000015",
        user_id: "00000000-0000-0000-0000-000000000010"
      }
    ]);
    client.mockReturnValueOnce([
      { id: "00000000-0000-0000-0000-000000000010" }
    ]);
    const { errors, data } = await addMessage.user({ input });
    expect(errors).toBeUndefined();
    expect(client).toMatchSnapshot();
    expect(client).toHaveBeenCalledTimes(4);
    expect(data).toEqual({
      addMessage: {
        status: "ok",
        message: {
          content: "hello world",
          user: { id: new User("00000000-0000-0000-0000-000000000010").id }
        }
      }
    });
  });

  it("when conversation not found", async () => {
    client.mockReturnValueOnce([]);
    const { errors } = await addMessage.user({ input });
    expect(errors[0].message).toBe("not found");
  });
});
