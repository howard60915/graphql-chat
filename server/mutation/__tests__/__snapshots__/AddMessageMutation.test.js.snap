// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMessageMutation successfully add message 1`] = `
Array [
  "SELECT * FROM \\"user_conversation\\" WHERE \\"user_id\\" = $1 AND \\"conversation_id\\" = $2 LIMIT 1",
  "SELECT * FROM \\"conversation\\" WHERE \\"conversation\\".\\"deleted_at\\" IS NULL AND \\"id\\" IN ($1)",
  "INSERT INTO \\"message\\" (\\"user_id\\",\\"content\\",\\"conversation_id\\",\\"created_by\\",\\"updated_by\\",\\"created_at\\",\\"updated_at\\") VALUES ($1,$2,$3,$4,$5,$6,$7) RETURNING *",
  "SELECT * FROM \\"user\\" WHERE \\"user\\".\\"deleted_at\\" IS NULL AND \\"id\\" IN ($1)",
]
`;
